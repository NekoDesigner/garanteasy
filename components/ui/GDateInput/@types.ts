/**
 * IGDateInputProps - Generated by Garanteasier CLI
 */

import { StyleProp, TextInputProps, TextStyle, ViewStyle } from "react-native";

export interface IGDateInputProps extends TextInputProps {
  style?: object;
  testID?: string;
  containerStyle?: StyleProp<ViewStyle>;
  labelStyle?: StyleProp<TextStyle>;
  errorStyle?: StyleProp<TextStyle>;
  label: string;
  value?: string; // Default value in DD/MM/YYYY format
  allowFutureDates?: boolean;
  onDateValidation?: (isValid: boolean, errorMessage?: string) => void;
  onDateChange?: (date: Date) => void;
}