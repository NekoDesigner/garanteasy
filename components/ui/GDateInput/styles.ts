/**
 * GDateInputStyles - Generated by Garanteasier CLI
 */

import { StyleSheet } from 'react-native';
import { COLORS } from '../../../constants';

const styles = StyleSheet.create({
  label: {
      marginBottom: 5,
      color: COLORS.placeholder,
      fontSize: 14,
    },
    textInput: {
      borderWidth: 1,
      borderColor: '#ccc',
      borderRadius: 4,
      padding: 10,
      fontSize: 16,
      color: '#333',
      backgroundColor: '#fff',
    },
    textInputError: {
      borderColor: '#ff4444',
    },
    errorText: {
      color: '#ff4444',
      fontSize: 12,
      marginTop: 5,
    }
});

export default styles;
