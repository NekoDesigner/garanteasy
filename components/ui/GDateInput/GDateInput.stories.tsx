/**
 * GDateInput - Generated by Garanteasier CLI
 */

import React from "react";
import { View } from "react-native";
import GDateInput from "./index";

export default {
  title: 'GDateInput',
  component: GDateInput
};

export const Default = () => (
    <View style={{ flex: 1, justifyContent: 'center', alignItems: 'center', padding: 16, gap: 24 }}>
      <GDateInput />
    </View>
);