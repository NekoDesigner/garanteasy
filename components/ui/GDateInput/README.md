# GDateInput

Un composant de saisie de date intelligent pour React Native avec formatage automatique et validation.

## Fonctionnalités

- ✅ **Formatage automatique** : L'utilisateur tape seulement les chiffres, le composant ajoute automatiquement les "/" pour le format DD/MM/YYYY
- ✅ **Validation de date** : Vérifie que la date existe dans le calendrier
- ✅ **Contrôle des dates futures** : Option pour autoriser ou interdire les dates supérieures à aujourd'hui
- ✅ **Messages d'erreur** : Affichage d'erreurs contextuelles lors de la validation
- ✅ **Feedback visuel** : Bordure rouge en cas d'erreur

## Props

| Prop | Type | Défaut | Description |
|------|------|--------|-------------|
| `label` | `string` | - | **Requis.** Libellé du champ |
| `allowFutureDates` | `boolean` | `false` | Autorise les dates futures |
| `onDateValidation` | `(isValid: boolean, errorMessage?: string) => void` | - | Callback appelé lors de la validation |
| `errorStyle` | `StyleProp<TextStyle>` | - | Style personnalisé pour le message d'erreur |
| `labelStyle` | `StyleProp<TextStyle>` | - | Style personnalisé pour le libellé |
| `containerStyle` | `StyleProp<ViewStyle>` | - | Style personnalisé pour le conteneur |

*+ toutes les props de `TextInput` de React Native*

## Utilisation

### Exemple basique

```tsx
import GDateInput from './components/ui/GDateInput';

function MyComponent() {
  const [date, setDate] = useState('');

  return (
    <GDateInput
      label="Date de naissance"
      value={date}
      onChangeText={setDate}
      allowFutureDates={false}
    />
  );
}
```

### Avec validation personnalisée

```tsx
import GDateInput from './components/ui/GDateInput';

function MyComponent() {
  const [date, setDate] = useState('');

  const handleValidation = (isValid: boolean, errorMessage?: string) => {
    if (!isValid) {
      console.log('Erreur:', errorMessage);
      // Gérer l'erreur (ex: afficher un toast, etc.)
    }
  };

  return (
    <GDateInput
      label="Date d'événement"
      value={date}
      onChangeText={setDate}
      allowFutureDates={true}
      onDateValidation={handleValidation}
    />
  );
}
```

## Comportement

### Formatage automatique
- L'utilisateur tape `15062025`
- Le composant affiche automatiquement `15/06/2025`

### Validation
La validation se déclenche lors de la perte de focus (`onBlur`) et vérifie :

1. **Format** : DD/MM/YYYY (2 chiffres jour, 2 chiffres mois, 4 chiffres année)
2. **Existence** : La date doit exister dans le calendrier (ex: 31/02/2025 est invalide)
3. **Dates futures** : Si `allowFutureDates={false}`, les dates supérieures à aujourd'hui sont rejetées

### Messages d'erreur
- `"Format invalide. Utilisez DD/MM/YYYY"` : Format incorrect
- `"Date invalide"` : Date inexistante dans le calendrier
- `"Date future non autorisée"` : Date future quand `allowFutureDates={false}`

## Styles

Le composant utilise des styles prédéfinis mais accepte des styles personnalisés :

```tsx
<GDateInput
  label="Ma date"
  containerStyle={{ marginBottom: 20 }}
  labelStyle={{ color: 'blue' }}
  errorStyle={{ fontSize: 14, fontWeight: 'bold' }}
  style={{ borderColor: 'green' }}
/>
```

## Accessibilité

- `testID` par défaut : `'gdateinput'`
- `testID` pour l'input : `'gdateinput-input'`
- `testID` pour l'erreur : `'gdateinput-error'`
