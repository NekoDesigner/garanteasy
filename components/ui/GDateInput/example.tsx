/**
 * Example d'utilisation du composant GDateInput
 */

import React, { useState } from 'react';
import { View, Text, StyleSheet } from 'react-native';
import GDateInput from './index';

const GDateInputExample = () => {
  const [birthDate, setBirthDate] = useState('');
  const [eventDate, setEventDate] = useState('');

  const handleBirthDateValidation = (isValid: boolean, errorMessage?: string) => {
    if (!isValid && errorMessage) {
      console.log('Erreur date de naissance:', errorMessage);
    }
  };

  const handleEventDateValidation = (isValid: boolean, errorMessage?: string) => {
    if (!isValid && errorMessage) {
      console.log('Erreur date événement:', errorMessage);
    }
  };

  return (
    <View style={styles.container}>
      <Text style={styles.title}>Exemples GDateInput</Text>

      {/* Date de naissance - pas de dates futures */}
      <GDateInput
        label="Date de naissance"
        value={birthDate}
        onChangeText={setBirthDate}
        allowFutureDates={false}
        onDateValidation={handleBirthDateValidation}
        style={styles.input}
      />

      {/* Date d'événement - dates futures autorisées */}
      <GDateInput
        label="Date d'événement"
        value={eventDate}
        onChangeText={setEventDate}
        allowFutureDates={true}
        onDateValidation={handleEventDateValidation}
        style={styles.input}
      />

      <Text style={styles.info}>
        • Tapez uniquement les chiffres (ex: 15062025)
        • Le format DD/MM/YYYY est appliqué automatiquement
        • La validation se fait lors de la perte de focus
      </Text>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    padding: 20,
    backgroundColor: '#f5f5f5',
  },
  title: {
    fontSize: 20,
    fontWeight: 'bold',
    marginBottom: 20,
    textAlign: 'center',
  },
  input: {
    marginBottom: 20,
  },
  info: {
    fontSize: 12,
    color: '#666',
    fontStyle: 'italic',
    marginTop: 20,
  },
});

export default GDateInputExample;
