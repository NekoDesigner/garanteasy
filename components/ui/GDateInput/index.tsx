/**
 * GDateInput - Generated by Garanteasier CLI
 */

import React, { useState, useRef } from "react";
import { View, Text, TextInput } from "react-native";
import CalendarIcon from "../Icons/CalendarIcon";
import { IGDateInputProps } from "./@types";
import GDateInputStyles from "./styles";

const GDateInput: React.FC<IGDateInputProps> = ({
  style,
  testID = 'gdateinput',
  label,
  labelStyle,
  errorStyle,
  value: initialValue,
  allowFutureDates = true,
  onDateValidation,
  onDateChange,
  onBlur,
  onChangeText, // Extract this to avoid conflicts
  ...props
}) => {
  const [value, setValue] = useState(initialValue || '');
  const [error, setError] = useState<string>('');
  const inputRef = useRef<TextInput>(null);

  // Update internal value when prop changes
  React.useEffect(() => {
    if (initialValue !== undefined) {
      setValue(initialValue);
    }
  }, [initialValue]);

  // Fonction pour formater la date en DD/MM/YYYY
  const formatDate = (text: string, previousValue: string = '') => {
    // Si l'utilisateur supprime tout, retourner une chaîne vide
    if (text === '') {
      return '';
    }

    // Supprimer tous les caractères non-numériques
    const cleanedText = text.replace(/\D/g, '');
    const previousCleanedText = previousValue.replace(/\D/g, '');

    // Si l'utilisateur supprime des caractères numériques, permettre la suppression
    if (cleanedText.length < previousCleanedText.length) {
      // L'utilisateur supprime des caractères, donc formatons juste ce qui reste
      const limitedText = cleanedText.slice(0, 8);

      let formattedDate = '';
      if (limitedText.length >= 1) {
        formattedDate += limitedText.slice(0, 2);
      }
      if (limitedText.length >= 3) {
        formattedDate += '/' + limitedText.slice(2, 4);
      }
      if (limitedText.length >= 5) {
        formattedDate += '/' + limitedText.slice(4, 8);
      }

      return formattedDate;
    }

    // Limiter à 8 chiffres maximum
    const limitedText = cleanedText.slice(0, 8);

    // Formater en DD/MM/YYYY
    let formattedDate = '';
    if (limitedText.length >= 1) {
      formattedDate += limitedText.slice(0, 2);
    }
    if (limitedText.length >= 3) {
      formattedDate += '/' + limitedText.slice(2, 4);
    }
    if (limitedText.length >= 5) {
      formattedDate += '/' + limitedText.slice(4, 8);
    }

    return formattedDate;
  };

  // Fonction pour valider si une date existe dans le calendrier
  const isValidDate = (day: number, month: number, year: number): boolean => {
    if (day < 1 || day > 31 || month < 1 || month > 12 || year < 1900) {
      return false;
    }

    const date = new Date(year, month - 1, day);
    return date.getFullYear() === year &&
           date.getMonth() === month - 1 &&
           date.getDate() === day;
  };

  // Fonction pour valider la date complète
  const validateDate = (dateString: string): { isValid: boolean; errorMessage?: string } => {
    // Vérifier le format DD/MM/YYYY
    const dateRegex = /^(\d{2})\/(\d{2})\/(\d{4})$/;
    const match = dateString.match(dateRegex);

    if (!match) {
      return { isValid: false, errorMessage: 'Format invalide. Utilisez DD/MM/YYYY' };
    }

    const day = parseInt(match[1], 10);
    const month = parseInt(match[2], 10);
    const year = parseInt(match[3], 10);

    // Vérifier si la date existe
    if (!isValidDate(day, month, year)) {
      return { isValid: false, errorMessage: 'Date invalide' };
    }

    // Vérifier si la date est dans le futur si non autorisé
    if (!allowFutureDates) {
      const inputDate = new Date(year, month - 1, day);
      const today = new Date();
      today.setHours(0, 0, 0, 0); // Réinitialiser l'heure pour comparer seulement la date

      if (inputDate > today) {
        return { isValid: false, errorMessage: 'Date future non autorisée' };
      }
    }

    return { isValid: true };
  };

  // Gérer le changement de texte
  const handleChangeText = (text: string) => {
    const formattedText = formatDate(text, value);
    setValue(formattedText);

    // Effacer l'erreur lors de la saisie
    if (error) {
      setError('');
    }

    // Appeler la fonction onChangeText si elle existe
    if (onChangeText) {
      onChangeText(formattedText);
    }

    if (onDateChange) {
      // Extraire le jour, le mois et l'année du texte formaté
      const parts = formattedText.split('/');
      if (parts.length === 3 && parts[0].length === 2 && parts[1].length === 2 && parts[2].length === 4) {
        const day = parseInt(parts[0], 10);
        const month = parseInt(parts[1], 10);
        const year = parseInt(parts[2], 10);

        // Vérifier si la date est valide avant de l'envoyer
        if (isValidDate(day, month, year)) {
          onDateChange(new Date(year, month - 1, day));
        }
      }
    }
  };

  // Gérer la perte de focus
  const handleBlur = (event: any) => {
    const validation = validateDate(value);

    if (!validation.isValid && value.length > 0) {
      setError(validation.errorMessage || '');
    } else {
      setError('');
    }

    // Appeler le callback de validation si fourni
    if (onDateValidation) {
      onDateValidation(validation.isValid, validation.errorMessage);
    }

    // Appeler le onBlur original si fourni
    if (onBlur) {
      onBlur(event);
    }
  };

  return (
    <View style={[GDateInputStyles, style]} testID={testID}>
      <Text style={[GDateInputStyles.label, labelStyle]}>{label}</Text>
      <View style={GDateInputStyles.textInputContainer} testID={`${testID}-container`}>
        <TextInput
          ref={inputRef}
          placeholder="DD/MM/YYYY"
          autoCapitalize="none"
          autoCorrect={false}
          testID={`${testID}-input`}
          {...props}
          value={value}
          keyboardType="numeric"
          maxLength={10} // Format DD/MM/YYYY
          style={[
            GDateInputStyles.textInput,
            error ? GDateInputStyles.textInputError : null,
            style
          ]}
          onChangeText={handleChangeText}
          onBlur={handleBlur}
        />
        <CalendarIcon />
      </View>
      {error ? (
        <Text style={[GDateInputStyles.errorText, errorStyle]} testID={`${testID}-error`}>
          {error}
        </Text>
      ) : null}
    </View>
  );
};

export default GDateInput;
