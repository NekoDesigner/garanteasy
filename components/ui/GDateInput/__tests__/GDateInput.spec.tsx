/**
 * GDateInput - Generated by Garanteasier CLI
 */

import { render } from "@testing-library/react-native";
import React from "react";
import GDateInput from "../index";

describe("GDateInput Components", () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

    test('renders correctly', () => {
      const tree = render(<GDateInput label="Start Date" />).toJSON();
      expect(tree).toMatchSnapshot();
    });
});