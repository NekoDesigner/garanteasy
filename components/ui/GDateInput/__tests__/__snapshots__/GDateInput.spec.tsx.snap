// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GDateInput Components renders correctly 1`] = `
<View
  style={
    [
      {
        "errorText": {
          "color": "#ff4444",
          "fontSize": 12,
          "marginTop": 5,
        },
        "label": {
          "color": "#454545",
          "fontSize": 14,
          "marginBottom": 5,
        },
        "textInput": {
          "backgroundColor": "#fff",
          "borderColor": "#ccc",
          "borderRadius": 4,
          "borderWidth": 1,
          "color": "#333",
          "fontSize": 16,
          "padding": 10,
        },
        "textInputError": {
          "borderColor": "#ff4444",
        },
      },
      undefined,
    ]
  }
  testID="gdateinput"
>
  <Text
    style={
      [
        {
          "color": "#454545",
          "fontSize": 14,
          "marginBottom": 5,
        },
        undefined,
      ]
    }
  />
  <TextInput
    autoCapitalize="none"
    autoCorrect={false}
    keyboardType="numeric"
    maxLength={10}
    onBlur={[Function]}
    onChangeText={[Function]}
    placeholder="DD/MM/YYYY"
    style={
      [
        {
          "backgroundColor": "#fff",
          "borderColor": "#ccc",
          "borderRadius": 4,
          "borderWidth": 1,
          "color": "#333",
          "fontSize": 16,
          "padding": 10,
        },
        null,
        undefined,
      ]
    }
    testID="gdateinput-input"
    value=""
  />
</View>
`;
